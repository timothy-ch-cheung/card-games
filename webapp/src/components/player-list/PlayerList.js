import React from 'react';
import Card from "react-bootstrap/Card";
import Button from "react-bootstrap/Button";
import Table from "react-bootstrap/Table";
import {isBlank} from "../../Util";

function Player(props) {
    let text = props.isHost ? props.name + " ðŸ‘‘" : props.name;

    return (
        <p style={{marginBottom: "0"}}>{text}</p>
    );

    return null;
}


function PlayerList(props) {

    const renderPlayer = (card, index) => {
        if (!isBlank(card.name)) {
            return (
                <tr>
                    <td style={{padding: "5px"}}><Player name={card.name} isHost={card.isHost} key={index}/></td>
                </tr>);
        }
        return null;
    }

    return (
        <Card style={{width: "35%", margin: "10px"}}>
            <Card.Header>Lobby</Card.Header>
            <Card.Body style={{padding: "5px 5px", height: "400px", overflowY: "auto"}}>
                <Table borderless variant="light">
                    <tbody>
                    {props.players.map(renderPlayer)}
                    </tbody>
                </Table>
            </Card.Body>
            <Card.Body style={{borderTop: "1px solid #dfdfdf", padding: "10px"}}>
                <Button variant="info" className="leave-game" onClick={props.onLeave}>Leave</Button>
            </Card.Body>

        </Card>
    );
};

export default PlayerList;