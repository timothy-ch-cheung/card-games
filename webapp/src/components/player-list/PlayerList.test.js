import React from 'react';
import {configure, mount, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import PlayerList from "./PlayerList";
import each from "jest-each";

configure({adapter: new Adapter()});

describe("TEST SUITE PlayerList: ", () => {
    const players = [{username: "John", isHost: true}, {username: "Jane", isHost: false}];

    test('Matches PlayerList snapshot', () => {
        const wrapper = mount(<PlayerList players={players}/>);
        expect(wrapper).toMatchSnapshot();
    });

    test('Player name has crown emoji if they are host', () => {
        const wrapper = shallow(<PlayerList players={[{username: "John", isHost: true}]}/>);
        expect(wrapper.find('[data-test="player0-row"]').dive()
            .find('[data-test="player0-name"]').text()).toEqual('John ðŸ‘‘')
    });

    test('Player name does not have crown emoji if they are guest', () => {
        const wrapper = shallow(<PlayerList players={[{username: "Jane", isHost: false}]}/>);
        expect(wrapper.find('[data-test="player0-row"]').dive()
            .find('[data-test="player0-name"]').text()).toEqual('Jane')
    });

    each([undefined, null, "", "      "]).test('Row is not rendered if player name is blank', (blankValue) => {
        const wrapper = shallow(<PlayerList
            players={[{username: "Jack", isHost: true}, {username: blankValue, isHost: false}]}/>);
        expect(wrapper.find('[data-test="player0-row"]').length).toEqual(1);
        expect(wrapper.find('[data-test="player1-row"]').length).toEqual(0);
    });

    test('number of players in lobby is shown', () => {
        const wrapper = shallow(<PlayerList maxPlayers={6}
            players={[{username: "Jack", isHost: true}, {username: "Jane", isHost: false}]}/>);
        expect(wrapper.find('p[data-test="player-count"]').text()).toEqual("2/6")
    });
});