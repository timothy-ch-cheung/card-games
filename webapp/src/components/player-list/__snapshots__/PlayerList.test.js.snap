// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST SUITE PlayerList:  Matches PlayerList snapshot 1`] = `
<PlayerList
  players={
    Array [
      Object {
        "isHost": true,
        "name": "John",
      },
      Object {
        "isHost": false,
        "name": "Jane",
      },
    ]
  }
>
  <Card
    body={false}
    style={
      Object {
        "margin": "10px",
        "width": "35%",
      }
    }
  >
    <div
      className="card"
      style={
        Object {
          "margin": "10px",
          "width": "35%",
        }
      }
    >
      <CardHeader>
        <div
          className="card-header"
        >
          Lobby
        </div>
      </CardHeader>
      <CardBody
        style={
          Object {
            "height": "400px",
            "overflowY": "auto",
            "padding": "5px 5px",
          }
        }
      >
        <div
          className="card-body"
          style={
            Object {
              "height": "400px",
              "overflowY": "auto",
              "padding": "5px 5px",
            }
          }
        >
          <ForwardRef
            borderless={true}
            variant="light"
          >
            <table
              className="table table-light table-borderless"
            >
              <tbody>
                <tr
                  key="0"
                >
                  <td
                    style={
                      Object {
                        "padding": "5px",
                      }
                    }
                  >
                    <Player
                      data-test="player0-name"
                      isHost={true}
                      key="0"
                      name="John"
                    >
                      <p
                        style={
                          Object {
                            "marginBottom": "0",
                          }
                        }
                      >
                        John ðŸ‘‘
                      </p>
                    </Player>
                  </td>
                </tr>
                <tr
                  key="1"
                >
                  <td
                    style={
                      Object {
                        "padding": "5px",
                      }
                    }
                  >
                    <Player
                      data-test="player1-name"
                      isHost={false}
                      key="1"
                      name="Jane"
                    >
                      <p
                        style={
                          Object {
                            "marginBottom": "0",
                          }
                        }
                      >
                        Jane
                      </p>
                    </Player>
                  </td>
                </tr>
              </tbody>
            </table>
          </ForwardRef>
        </div>
      </CardBody>
      <CardBody
        style={
          Object {
            "borderTop": "1px solid #dfdfdf",
            "padding": "10px",
          }
        }
      >
        <div
          className="card-body"
          style={
            Object {
              "borderTop": "1px solid #dfdfdf",
              "padding": "10px",
            }
          }
        >
          <Button
            active={false}
            className="leave-game"
            disabled={false}
            type="button"
            variant="info"
          >
            <button
              className="leave-game btn btn-info"
              disabled={false}
              type="button"
            >
              Leave
            </button>
          </Button>
        </div>
      </CardBody>
    </div>
  </Card>
</PlayerList>
`;
